File: freelance_AI_back-main/internal/domain/notification/catalog.go
package notification

import (
	"encoding/json"
	"strings"
)

type NormalizedFields struct {
	Type    string
	Title   string
	Message string
	Link    string
	Event   string
}

type fallback struct {
	Type    string
	Title   string
	Message string
}

var eventFallbacks = map[string]fallback{
	"notification.created": {
		Type:    "system",
		Title:   "Новое уведомление",
		Message: "Проверьте центр уведомлений",
	},
	"proposal.created": {
		Type:    "proposal",
		Title:   "Новый отклик",
		Message: "Входящие отклики обновлены",
	},
	"proposal.updated": {
		Type:    "proposal",
		Title:   "Отклик обновлен",
		Message: "Статус отклика изменился",
	},
	"order.created": {
		Type:    "order",
		Title:   "Новый заказ",
		Message: "Лента заказов обновлена",
	},
	"order.updated": {
		Type:    "order",
		Title:   "Заказ обновлен",
		Message: "Изменения в карточке заказа",
	},
	"chat.message.created": {
		Type:    "message",
		Title:   "Новое сообщение",
		Message: "Откройте чаты, чтобы ответить",
	},
	"chat.message.updated": {
		Type:    "message",
		Title:   "Сообщение обновлено",
		Message: "Содержимое сообщения изменено",
	},
	"balance.updated": {
		Type:    "payment",
		Title:   "Баланс обновлен",
		Message: "Проверьте актуальное состояние баланса",
	},
	"transaction.created": {
		Type:    "payment",
		Title:   "Новая транзакция",
		Message: "История операций обновлена",
	},
	"proposal.ai_analysis_ready": {
		Type:    "ai",
		Title:   "AI-анализ готов",
		Message: "Анализ откликов завершён",
	},
	"session.revoked": {
		Type:    "system",
		Title:   "Сессия завершена",
		Message: "Выполнен выход из аккаунта",
	},
}

var typeFallbacks = map[string]fallback{
	"message": {
		Title:   "Сообщение",
		Message: "Откройте сообщения для деталей",
	},
	"proposal": {
		Title:   "Обновление отклика",
		Message: "Проверьте изменения по отклику",
	},
	"order": {
		Title:   "Обновление заказа",
		Message: "Проверьте изменения по заказу",
	},
	"payment": {
		Title:   "Финансовое обновление",
		Message: "Проверьте историю операций",
	},
	"ai": {
		Title:   "AI-обновление",
		Message: "Новая AI-активность по вашему аккаунту",
	},
	"system": {
		Title:   "Уведомление",
		Message: "Проверьте центр уведомлений",
	},
}

func NormalizePayload(payload json.RawMessage) NormalizedFields {
	root := toObject(payload)
	nested := toObject(root["data"])

	eventName := firstString(
		root["event"],
		root["event_type"],
		root["kind"],
		nested["event"],
		nested["event_type"],
		nested["kind"],
	)
	rawType := firstString(
		root["type"],
		root["category"],
		nested["type"],
		nested["category"],
	)

	normalizedType := normalizeType(rawType, eventName)
	fallbackByEvent, hasEventFallback := eventFallbacks[eventName]
	fallbackByType, hasTypeFallback := typeFallbacks[normalizedType]

	resolvedType := normalizedType
	if hasEventFallback && fallbackByEvent.Type != "" {
		resolvedType = fallbackByEvent.Type
	}
	if resolvedType == "" {
		resolvedType = "system"
	}

	title := firstString(root["title"], root["name"], nested["title"], nested["name"])
	if title == "" {
		if hasEventFallback {
			title = fallbackByEvent.Title
		} else if hasTypeFallback {
			title = fallbackByType.Title
		} else {
			title = "Уведомление"
		}
	}

	message := firstString(
		root["message"],
		root["description"],
		root["body"],
		root["text"],
		nested["message"],
		nested["description"],
		nested["body"],
		nested["text"],
	)
	if message == "" {
		if hasEventFallback {
			message = fallbackByEvent.Message
		} else if hasTypeFallback {
			message = fallbackByType.Message
		} else {
			message = "Проверьте центр уведомлений"
		}
	}

	link := firstString(root["link"], root["url"], root["path"], nested["link"], nested["url"], nested["path"])

	return NormalizedFields{
		Type:    resolvedType,
		Title:   title,
		Message: message,
		Link:    link,
		Event:   eventName,
	}
}

func toObject(raw interface{}) map[string]interface{} {
	if raw == nil {
		return map[string]interface{}{}
	}
	if bytes, ok := raw.(json.RawMessage); ok {
		var decoded map[string]interface{}
		if err := json.Unmarshal(bytes, &decoded); err == nil {
			return decoded
		}
		return map[string]interface{}{}
	}
	if object, ok := raw.(map[string]interface{}); ok {
		return object
	}
	if bytes, ok := raw.([]byte); ok {
		var decoded map[string]interface{}
		if err := json.Unmarshal(bytes, &decoded); err == nil {
			return decoded
		}
		return map[string]interface{}{}
	}
	if text, ok := raw.(string); ok {
		var decoded map[string]interface{}
		if err := json.Unmarshal([]byte(text), &decoded); err == nil {
			return decoded
		}
		return map[string]interface{}{}
	}
	return map[string]interface{}{}
}

func firstString(values ...interface{}) string {
	for _, raw := range values {
		str, ok := raw.(string)
		if !ok {
			continue
		}
		trimmed := strings.TrimSpace(str)
		if trimmed != "" {
			return trimmed
		}
	}
	return ""
}

func normalizeType(rawType, eventName string) string {
	source := strings.ToLower(rawType + " " + eventName)
	switch {
	case strings.Contains(source, "proposal"):
		return "proposal"
	case strings.Contains(source, "order"):
		return "order"
	case strings.Contains(source, "chat.message"), strings.Contains(source, "message"):
		return "message"
	case strings.Contains(source, "balance"), strings.Contains(source, "transaction"), strings.Contains(source, "payment"):
		return "payment"
	case strings.Contains(source, "ai"):
		return "ai"
	default:
		if rawType != "" {
			return rawType
		}
		return "system"
	}
}

File: frontend/src/shared/notifications/catalog.ts
const asObject = (value: unknown): Record<string, unknown> => {
  if (value && typeof value === "object") return value as Record<string, unknown>;
  if (typeof value === "string") {
    try {
      const parsed = JSON.parse(value) as unknown;
      if (parsed && typeof parsed === "object") return parsed as Record<string, unknown>;
    } catch {
      return {};
    }
  }
  return {};
};

const pickString = (...values: unknown[]): string | undefined => {
  for (const value of values) {
    if (typeof value !== "string") continue;
    const trimmed = value.trim();
    if (trimmed) return trimmed;
  }
  return undefined;
};

export type NotificationKind = "message" | "proposal" | "order" | "payment" | "review" | "ai" | "system";

type NotificationFallback = {
  kind: NotificationKind;
  title: string;
  message: string;
};

const EVENT_DEFAULTS: Record<string, NotificationFallback> = {
  "notification.created": {
    kind: "system",
    title: "Новое уведомление",
    message: "Проверьте центр уведомлений",
  },
  "proposal.created": {
    kind: "proposal",
    title: "Новый отклик",
    message: "Входящие отклики обновлены",
  },
  "proposal.updated": {
    kind: "proposal",
    title: "Отклик обновлен",
    message: "Статус отклика изменился",
  },
  "order.created": {
    kind: "order",
    title: "Новый заказ",
    message: "Лента заказов обновлена",
  },
  "order.updated": {
    kind: "order",
    title: "Заказ обновлен",
    message: "Изменения в карточке заказа",
  },
  "order.application_closed": {
    kind: "order",
    title: "Приём откликов завершён",
    message: "Публикация заказа обновлена",
  },
  "chat.message.created": {
    kind: "message",
    title: "Новое сообщение",
    message: "Откройте чаты, чтобы ответить",
  },
  "chat.message.updated": {
    kind: "message",
    title: "Сообщение обновлено",
    message: "Содержимое сообщения изменено",
  },
  "balance.updated": {
    kind: "payment",
    title: "Баланс обновлен",
    message: "Проверьте актуальное состояние баланса",
  },
  "transaction.created": {
    kind: "payment",
    title: "Новая транзакция",
    message: "История операций обновлена",
  },
  "proposal.ai_analysis_ready": {
    kind: "ai",
    title: "AI-анализ готов",
    message: "Анализ откликов завершён",
  },
  "session.revoked": {
    kind: "system",
    title: "Сессия завершена",
    message: "Выполнен выход из аккаунта",
  },
};

const KIND_DEFAULTS: Record<NotificationKind, Pick<NotificationFallback, "title" | "message">> = {
  message: {
    title: "Сообщение",
    message: "Откройте сообщения для деталей",
  },
  proposal: {
    title: "Обновление отклика",
    message: "Проверьте изменения по отклику",
  },
  order: {
    title: "Обновление заказа",
    message: "Проверьте изменения по заказу",
  },
  payment: {
    title: "Финансовое обновление",
    message: "Проверьте историю операций",
  },
  review: {
    title: "Обновление отзыва",
    message: "Появились изменения по отзывам",
  },
  ai: {
    title: "AI-обновление",
    message: "Новая AI-активность по вашему аккаунту",
  },
  system: {
    title: "Уведомление",
    message: "Проверьте центр уведомлений",
  },
};

const normalizeKind = (rawType: string, eventName: string): NotificationKind => {
  const source = `${rawType} ${eventName}`.toLowerCase();
  if (source.includes("proposal")) return "proposal";
  if (source.includes("order")) return "order";
  if (source.includes("chat.message") || source.includes("message")) return "message";
  if (source.includes("balance") || source.includes("transaction") || source.includes("payment")) return "payment";
  if (source.includes("review")) return "review";
  if (source.includes("ai")) return "ai";
  return "system";
};

export type UiNotification = {
  kind: NotificationKind;
  eventName: string;
  title: string;
  message: string;
  link?: string;
};

type MapInput = {
  eventName?: string;
  rawType?: string;
  title?: unknown;
  message?: unknown;
  link?: unknown;
  payload?: unknown;
};

export const mapEventToUiNotification = (input: MapInput): UiNotification => {
  const payload = asObject(input.payload);
  const nested = asObject(payload.data ?? payload.notification ?? payload.content ?? payload.meta);

  const eventName =
    pickString(
      input.eventName,
      payload.event,
      payload.event_type,
      payload.kind,
      nested.event,
      nested.event_type,
      nested.kind,
    ) ?? "";
  const rawType = pickString(input.rawType, payload.type, payload.category, nested.type, nested.category) ?? "";
  const kind = normalizeKind(rawType, eventName);
  const eventFallback = EVENT_DEFAULTS[eventName];
  const kindFallback = KIND_DEFAULTS[kind];

  const title =
    pickString(input.title, payload.title, payload.name, nested.title, nested.name) ??
    eventFallback?.title ??
    kindFallback.title;
  const message =
    pickString(
      input.message,
      payload.message,
      payload.description,
      payload.body,
      payload.text,
      nested.message,
      nested.description,
      nested.body,
      nested.text,
    ) ??
    eventFallback?.message ??
    kindFallback.message;
  const link = pickString(input.link, payload.link, payload.url, payload.path, nested.link, nested.url, nested.path);

  return {
    kind: eventFallback?.kind ?? kind,
    eventName,
    title,
    message,
    ...(link ? { link } : {}),
  };
};

export const isTechnicalNotificationEvent = (eventName: string): boolean =>
  eventName === "connection.state" || eventName === "notification.unread_count.updated";

